File: frontend/src/lib/translator.ts
"use client";

const cache = new Map<string, string>();

function key(to: string, text: string): string {
  return `${to}\u0001${text}`;
}

export interface Translator {
  translate: (text: string, to: string) => Promise<string>;
  translateBatch: (texts: string[], to: string) => Promise<string[]>;
}

/**
 * Client-side Azure Translator wrapper. Translates via the server proxy
 * (/api/translate) and caches results in-memory. On any failure, returns the
 * original text so the UI still renders (English fallback).
 */
export function createTranslator(): Translator {
  async function call(texts: string[], to: string): Promise<string[]> {
    const res = await fetch("/api/translate", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ texts, to }),
    });
    if (!res.ok) throw new Error(`translate ${res.status}`);
    const data = (await res.json()) as { translations: string[] };
    return data.translations;
  }

  async function lookup(text: string, to: string): Promise<string> {
    const k = key(to, text);
    const hit = cache.get(k);
    if (hit !== undefined) return hit;
    try {
      const [translated] = await call([text], to);
      const out = translated || text;
      cache.set(k, out);
      return out;
    } catch {
      return text;
    }
  }

  async function batch(texts: string[], to: string): Promise<string[]> {
    const missing: number[] = [];
    const out = new Array<string>(texts.length);
    texts.forEach((t, i) => {
      const hit = cache.get(key(to, t));
      if (hit !== undefined) out[i] = hit;
      else missing.push(i);
    });
    if (missing.length) {
      try {
        const translated = await call(missing.map((i) => texts[i]), to);
        missing.forEach((i, k) => {
          const v = translated[k] || texts[i];
          out[i] = v;
          cache.set(key(to, texts[i]), v);
        });
      } catch {
        missing.forEach((i) => {
          out[i] = texts[i];
        });
      }
    }
    return out;
  }

  return { translate: lookup, translateBatch: batch };
}

File: backend/app/providers/translator.py
"""Azure Translator provider for multilingual query normalization (Phase 10).

The authoritative corpus is English, so a non-English user query is translated
to English *before* embedding / classification / lexical retrieval. The
original query language is preserved for answer generation (the LLM answers in
the user's language). Translating only the retrieval-side query — never the
documents — keeps the frozen corpus untouched and avoids re-ingestion.

All failures degrade gracefully: if the translator is unconfigured or an
HTTP/parse error occurs, the original text is returned so retrieval still runs
(weaker, but never crashing).
"""
from __future__ import annotations

from functools import lru_cache

import httpx

from app.config import REQUEST_TIMEOUT_S, get_settings


class TranslatorError(RuntimeError):
    pass


@lru_cache(maxsize=1024)
def _cached_translate(key: str, text: str, to: str, source: str) -> str:
    """Translate with a stable cache key; returns original text on any failure.

    ``key`` is the translator subscription key so different deployments don't
    share cache entries; it is never logged or returned.
    """
    if not text.strip():
        return text
    endpoint = get_settings().azure_translator_endpoint.rstrip("/")
    params = {"api-version": "3.0", "to": to}
    if source:
        params["from"] = source
    headers = {
        "Ocp-Apim-Subscription-Key": get_settings().azure_translator_key,
        "Ocp-Apim-Subscription-Region": get_settings().azure_translator_region,
        "Content-Type": "application/json",
    }
    try:
        with httpx.Client(timeout=REQUEST_TIMEOUT_S) as client:
            resp = client.post(
                f"{endpoint}/translate",
                params=params,
                headers=headers,
                json=[{"Text": text}],
            )
            resp.raise_for_status()
            data = resp.json()
            translated = data[0]["translations"][0]["text"]
            return translated if translated else text
    except (httpx.HTTPError, KeyError, IndexError, ValueError):
        return text


class AzureTranslator:
    """Thin Azure Translator client. Configured when key + endpoint are set."""

    def __init__(self, settings=None) -> None:
        self.settings = settings or get_settings()

    @property
    def configured(self) -> bool:
        return bool(self.settings.azure_translator_key
                    and self.settings.azure_translator_endpoint)

    def translate(self, text: str, to: str = "en", source: str | None = None) -> str:
        if not self.configured:
            return text
        return _cached_translate(self.settings.azure_translator_key, text, to, source or "")
